controller/backup_controller.go

```diff
@@ -499,6 +499,26 @@ func (bc *BackupController) reconcile(backupName string) (err error) {
 		return nil
 	}
 
+	defer func() {
+		// If the backup is in the final state, delete the snapshot if needed
+		if _, ok := backupInfo.Labels[types.RecurringJobLabel]; ok {
+			// leave the snapshot management for recurring jobs to the `SettingNameAutoCleanupRecurringJobBackupSnapshot` setting.
+			return
+		}
+
+		cleanupSnap, err := bc.ds.GetSettingAsBool(types.SettingNameAutoCleanupSnapshotAfterOnDemandBackupCompleted)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get the setting %v", types.SettingNameAutoCleanupSnapshotAfterOnDemandBackupCompleted)
+			return
+		}
+		if cleanupSnap && backup.Spec.SnapshotName != "" {
+			if err := bc.ds.DeleteSnapshot(backup.Spec.SnapshotName); err != nil {
+				log.WithError(err).Error("Failed to delete the snapshot")
+				return
+			}
```

Comment:
The `err` will not pass to the returned value (err) of the function. You can use 

```
			err = bc.ds.DeleteSnapshot(backup.Spec.SnapshotName)
                         if err != nil {
mantissahz marked this conversation as resolved.
				log.WithError(err).Error("Failed to delete the snapshot")
				return
			}
```

Reply:
Modified. Thanks.